### ISSUE TITLE
Research identifying pieces w/ CV

### ISSUE THREAD

**Issue by WebCoder49** [Brighton-FTC/FTC-2023#26]

**Comment by 12sliu** [comment]
https://docs.ftclib.org/ftclib/vision/high-goal-detection

```
UGBasicHighGoalPipeline pipeline = new UGBasicHighGoalPipeline();
camera.setPipeline(pipeline);

// Red Goal
if (pipeline.isRedVisible()) {
    Rect redRect = pipeline.getRedRect();
    Point centerOfRedGoal = pipeline.getCenterofRect(redRect);
    
    telemetry.addData("Red goal position",
                        centerOfRedGoal.toString());
}

// say we have a point called 'centerOfCamera'
double xDist = Math.abs(centerOfCamera.x - centerOfBlueGoal.x);
double yDist = Math.abs(centerOfCamera.y - centerOfBlueGoal.y);
double dist = Math.hypot(xDist, yDist);

// assume we have a method that turns the robot at some speed,
// but clamps it to [-1, 1]
while (true) {
    if (centerOfCamera.x < centerOfBlueGoal.x - TOLERANCE) {
        turn(dist);    // turning right is positive input
    } else if (centerOfCamera.x > centerOfBlueGoal.x + TOLERANCE) {
        turn(-dist);
    } else {
        turn(0);
        break;
    }
}
```

My idea is we steal this but replace the colour with whatever colour the pieces are. If this works, we should be able to identify if there is a piece on screen, turn to face it, and then move forward until distance sensors detect something. There's also something with pitch and yaw that I have no clue about in the link provided.

**Comment by 12sliu** [comment]
took the base tensorflow thing and made some modifications, should detect if there is bolt, bulb, or panel and then send telemetry message, now looking at custom tensorflow thing

**Comment by WebCoder49** [comment]
@12sliu , this diagram (Leconfield's House Robotics Conceptual Flowchart) may be a useful idea for custom sleeves using TFOD.

![image](https://user-images.githubusercontent.com/69071853/226118788-68aebacb-a894-4673-87a3-c9478072cff5.png)

**Comment by 12sliu** [comment]
stuck on how to install protobuf compiler, needs to be installed for installing tensorflow api

given the complexity of making a custom model, I think we should focus on getting what we have working first, then trying to train a tfod model

**Comment by WebCoder49** [comment]
I agree with you. Just in case you get onto the stage of using a custom model, [this may be useful in how to install TFOD](https://github.com/google/ftc-object-detection/blob/master/TFObjectDetector/README.md), and it doesn't look like you would need to install TensorFlow separately.

**Comment by 12sliu** [comment]
![Image](https://user-images.githubusercontent.com/118719719/226838164-2114b789-4845-43fe-baad-5380c173d25b.png)
cannot install tensorflow api, have asked mr tattam for help, has anyone encountered this before? 

removing use-feature from command results in this 
![Image](https://user-images.githubusercontent.com/118719719/226838613-6dfc3444-e4a1-4c2f-bec1-fe86c5b5607c.png)

am completely stuck, if nobody has a solution I'll post it on stack overflow

**Comment by WebCoder49** [comment]
If you use PyCharm with a `venv` interpreter then install the libraries in `Ctrl+Alt+S > Project > Python Interpreter` that may work.
![image](https://user-images.githubusercontent.com/69071853/227147642-20568545-ef0a-4e2c-93cb-249c047c9312.png)

**Comment by WebCoder49** [comment]
Also, you are using some bash commands like `cp`, etc. They don't work on Windows command.exe. However, if you open PowerShell from the start menu and use the same commands that may fix it.

**Comment by WebCoder49** [comment]
I'm afraid that I'm quite busy with FRC at the moment; I will address this as soon as possible.
I would recommend downloading the scripts from the repository and opening the `training` folder in PyCharm, then:
* Complete the instructions in the `README` file to save a dataset of objects from videos.

I wouldn't recommend installing the library from PIP, as it's likely to overcomplicate it. They are more utility scripts than a coherent module.

**Comment by 12sliu** [comment]
even if I pulled a all-nighter, we don't have a video of the custom sleeve, so can't begin to do things

will probably try to finish for proof of concept after tmw ftc done

**Comment by 12sliu** [comment]
waiting for sleeve to be made and video made

**Comment by sbottingota** [comment]
@12sliu, could you design and print out the custom sleeves for CV? (Order of Mo)

**Comment by WebCoder49** [comment]
> @12sliu, could you design and print out the custom sleeves for CV? (Order of Mo)

@12sliu this document may be useful if you have not seen it already: https://www.firstinspires.org/sites/default/files/uploads/resource_library/ftc/signal-sleeve-template.pdf

**Comment by 12sliu** [comment]
Thanks Oliver. 

Are we using April Tags or Tensorflow still? If Tensorflow, do I just stick whatever I want on the sleeve as long as the three sides are unique?

**Comment by 12sliu** [comment]
Never mind, apparently Benji from engineering has gotten the sleeve designed and done. Apparently there is some miscommunication.

**Comment by WebCoder49** [comment]
> Never mind, apparently Benji from engineering has gotten the sleeve designed and done. Apparently there is some miscommunication.

Yes; I think I just saw a Teams message from Benji.
